--- a/include/sc.hpp
+++ b/include/sc.hpp
@@ -117,4 +117,6 @@ class Sc : public Kobject, public Refcount
             time_m  = 0;
             tsc     = 0;
         }
+
+        uint16 vcpu_prio() const { return 15; }
 };
--- a/src/sc.cpp
+++ b/src/sc.cpp
@@ -70,6 +70,11 @@ void Sc::ready_enqueue (uint64 t, bool inc_ref, bool use_left)
             return;
     }
 
+#if 1
+    if (ec->vcpu() && prio != 15)
+        prio = 15;
+#endif
+
     if (prio > prio_top)
         prio_top = prio;
 
--- a/src/syscall.cpp
+++ b/src/syscall.cpp
@@ -56,6 +56,19 @@ void Ec::activate()
 {
     Ec *ec = this;
 
+    if (ec->vcpu()) {
+        if (ec->partner && Sc::current->prio != Sc::current->vcpu_prio() + 1) {
+            if (Sc::current->ec == ec)
+                Sc::current->prio = Sc::current->vcpu_prio() + 1;
+            else
+                trace(0, "%s %u vCPU    partner - %s", __func__, Sc::current->prio,
+                      Sc::current->ec == ec ? " right SC" : " wrong SC");
+        }
+        if (!ec->partner && Sc::current->prio != Sc::current->vcpu_prio())
+            trace(0, "%s %u vCPU no partner - %s", __func__, Sc::current->prio,
+                  Sc::current->ec == ec ? " right SC" : " wrong SC");
+    }
+
     // XXX: Make the loop preemptible
     for (Sc::ctr_link = 0; ec->partner; ec = ec->partner)
         Sc::ctr_link++;
@@ -111,6 +124,12 @@ void Ec::send_msg()
     if (EXPECT_TRUE (!ec->cont)) {
         current->cont = C;
         current->set_partner (ec);
+        if (current->vcpu() && Sc::current->prio == Sc::current->vcpu_prio()) {
+            Sc::current->prio = Sc::current->vcpu_prio() + 1;
+//            trace(0, "%s vCPU prio=%u", __func__, Sc::current->prio);
+            assert(Sc::current->ec == current);
+        }
+
         current->regs.mtd = pt->mtd.val;
         ec->cont = recv_kern;
         ec->regs.set_pt (pt->id);
@@ -154,6 +173,8 @@ void Ec::sys_call()
     if (EXPECT_TRUE (!ec->cont)) {
         current->cont = current->xcpu_sm ? xcpu_return : ret_user_sysexit;
         current->set_partner (ec);
+        if (current->vcpu())
+            trace(0, "%s vCPU prio=%u", __func__, Sc::current->prio);
         ec->cont = recv_user;
         ec->regs.set_pt (pt->id);
         ec->regs.set_ip (pt->ip);
@@ -218,6 +239,14 @@ void Ec::reply (void (*c)(), Sm * sm)
     if (sm)
         sm->dn (false, 0, ec, clr);
 
+    if (ec->vcpu() && Sc::current->ec == ec) {
+        if (Sc::current->prio != Sc::current->vcpu_prio()) {
+            Sc::current->prio = Sc::current->vcpu_prio();
+            Sc::schedule (false);
+        } else
+           trace(0, "%s already %u", __func__, Sc::current->vcpu_prio());
+    }
+
     if (!clr)
         Sc::current->ec->activate();
 
