diff --git a/src/syscall.cpp b/src/syscall.cpp
index 4a55821..e949a20 100644
--- a/src/syscall.cpp
+++ b/src/syscall.cpp
@@ -250,7 +253,22 @@ void Ec::sys_create_ec()
         sys_finish<Sys_regs::BAD_PAR>();
     }
 
-    Ec *ec = new Ec (Pd::current, r->sel(), pd, r->flags() & 1 ? static_cast<void (*)()>(send_msg<ret_user_iret>) : nullptr, r->cpu(), r->evt(), r->utcb(), r->esp());
+    void (*function) () = nullptr;
+    unsigned evt = r->evt();
+    if (r->flags() & 1)
+        function = static_cast<void (*)()>(send_msg<ret_user_iret>);
+ 
+    if (r->flags() & 2) {
+        function = ret_user_sysexit;
+        evt = 0;
+    }
+ 
+    Ec *ec = new Ec (Pd::current, r->sel(), pd, function, r->cpu(), evt, r->utcb(), r->esp());
+ 
+    if (r->flags() & 2) {
+        ec->regs.set_ip (r->evt());
+        ec->regs.set_sp (r->esp());
+    }
 
     if (!Space_obj::insert_root (ec)) {
         trace (TRACE_ERROR, "%s: Non-NULL CAP (%#lx)", __func__, r->sel());
