diff --git a/src/syscall.cpp b/src/syscall.cpp
index 5291815..2654686 100644
--- a/src/syscall.cpp
+++ b/src/syscall.cpp
@@ -477,19 +477,20 @@ void Ec::sys_assign_pci()
 
     Kobject *obj = Space_obj::lookup (r->pd()).obj();
     if (EXPECT_FALSE (obj->type() != Kobject::PD)) {
-        trace (TRACE_ERROR, "%s: Non-PD CAP (%#lx)", __func__, r->pd());
-        sys_finish<Sys_regs::BAD_CAP>();
+        trace (0, "%s: Non-PD CAP (%#lx)", __func__, r->pd());
+		obj = Pd::current;
+//        sys_finish<Sys_regs::BAD_CAP>();
     }
 
     Paddr phys; unsigned rid;
     if (EXPECT_FALSE (!Pd::current->Space_mem::lookup (r->dev(), phys) || (rid = Pci::phys_to_rid (phys)) == ~0U)) {
-        trace (TRACE_ERROR, "%s: Non-DEV CAP (%#lx)", __func__, r->dev());
+        trace (0, "%s: Non-DEV CAP (%#lx)", __func__, r->dev());
         sys_finish<Sys_regs::BAD_DEV>();
     }
 
     Dmar *dmar = Pci::find_dmar (r->hnt());
     if (EXPECT_FALSE (!dmar)) {
-        trace (TRACE_ERROR, "%s: Invalid Hint (%#lx)", __func__, r->hnt());
+        trace (0, "%s: Invalid Hint (%#lx)", __func__, r->hnt());
         sys_finish<Sys_regs::BAD_DEV>();
     }
 
