--- src/kernel/sel4/src/arch/x86/object/vcpu.c
+++ src/kernel/sel4/src/arch/x86/object/vcpu.c
@@ -1150,6 +1150,8 @@
          * that may be in the desired current cr0 */
         vcpu->cr0 = (vcpu->cached_cr0 & ~CR0_TASK_SWITCH) | (NODE_STATE(ksCurThread)->tcbArch.tcbVCPU->cr0 & CR0_TASK_SWITCH);
     }
+
+    switchFpuOwner(NULL, getCurrentCPUIndex());
 }
 
 exception_t handleVmexit(void)
@@ -1424,6 +1426,8 @@
         vmwrite(VMX_CONTROL_CR0_READ_SHADOW, cr0_shadow);
         vcpu->cached_cr0_shadow = cr0_shadow;
     }
+
+    switchFpuOwner(NULL, getCurrentCPUIndex());
 }
 
 void clearVPIDIOPortMappings(vpid_t vpid, uint16_t first, uint16_t last)
