--- a/src/app/virtualbox/src/VBox/Devices/Audio/HDAStreamPeriod.cpp
+++ b/src/app/virtualbox/src/VBox/Devices/Audio/HDAStreamPeriod.cpp
@@ -180,7 +180,9 @@
  */
 void hdaStreamPeriodEnd(PHDASTREAMPERIOD pPeriod)
 {
+#ifdef DEBUG
     Log3Func(("[SD%RU8] Took %zuus\n", pPeriod->u8SD, (RTTimeNanoTS() - pPeriod->Dbg.tsStartNs) / 1000));
+#endif
 
     if (!(pPeriod->fStatus & HDASTREAMPERIOD_FLAG_ACTIVE))
         return;
--- a/src/app/virtualbox/src/VBox/Devices/Audio/DevHDACommon.cpp
+++ b/src/app/virtualbox/src/VBox/Devices/Audio/DevHDACommon.cpp
@@ -64,7 +64,9 @@
     if (   (HDA_REG(pThis, INTCTL) & HDA_INTCTL_GIE)
         && (HDA_REG(pThis, INTSTS) & HDA_REG(pThis, INTCTL) & (HDA_INTCTL_CIE | HDA_STRMINT_MASK)))
     {
+#ifdef DEBUG
         Log3Func(("Asserted (%s)\n", pszSource));
+#endif
 
         PDMDevHlpPCISetIrq(pThis->CTX_SUFF(pDevIns), 0, 1 /* Assert */);
         pThis->u8IRQL = 1;
@@ -76,7 +78,9 @@
     }
     else
     {
+#ifdef DEBUG
         Log3Func(("Deasserted (%s)\n", pszSource));
+#endif
 
         PDMDevHlpPCISetIrq(pThis->CTX_SUFF(pDevIns), 0, 0 /* Deassert */);
         pThis->u8IRQL = 0;
